### ISSUE TITLE
TinyMCE in IonModal fails to start when initialised a second time

### ISSUE THREAD

**Issue by 1dl3** [tinymce/tinymce-react#204]
**What is the current behavior?**
I have an Editor on an IonModal if I open the modal twice the Editor crashes and fails to render.

**Please provide the steps to reproduce and if possible a minimal demo of the problem via [codesandbox.io](https://codesandbox.io/s/6yz4jzrrxw?fontsize=14) or similar.**
You can find a minimal example of this here: https://codesandbox.io/s/jolly-fire-d7l52?file=/src/App.js
If you press "Show Modal" twice it crashes with 

`tinymce.min.js:2 Uncaught TypeError: Cannot read property 'open' of undefined`

**What is the expected behavior?**

The Editor should show up. 

**Which versions of TinyMCE, and which browser / OS are affected by this issue? Did this work in previous versions of TinyMCE or `tinymce-react`?**

Google Chrome | 87.0.4280.141

@tinymce/tinymce-react 3.9.0
@ionic/react 5.5.2
@ionic/react-hooks 0.0.8
@ionic/react-route 5.3.4

I figure it's a problem with the IonModal and my current workaround to use the Material-UI Modal works. 
I just thought I should let someone know.

**Comment by tiny-james** [comment]
Well my best guess is that something to do with shadow-dom is interacting badly with TinyMCE because it seems to be having trouble getting things that are normally available globally on the window like document and styleSheets.

I tried it with the latest editor unminfied here:
https://codesandbox.io/s/wizardly-cherry-pl159?file=/src/App.js

Honestly I have no idea how to fix this.

**Comment by Marruixu** [comment]
Hi, I share a workaround that I have quickly come up with, which I hope it may help or inspire for those using Ionic 5 Modal with React:

- Use normal tinymce package, not wrapper tinymce-react, and in the module which holds the editor import it like: `import tinymce from 'tinymce/tinymce';`
- If you are using apiKey method, this is not required: At the end of <head> tag in index.html add `<script src="%PUBLIC_URL%/tinymce/tinymce.js"></script>` and place the unminfied (or minfied, then change URL accordingly) resources at /public/tinymce
- Set a `<textarea id={"myTinyMCE" + initCounter} defaultValue={content} />` as a IonModal child
- Define 2 states for initCounter and content (last one not really required) `const [initCounter, setInitCounter] = useState<number>(0);` and `const [content, setContent] = useState<string>();` (assuming Typescript here)
- Define useEffect based on IonModal isOpen property
```
useEffect(() => {
  if (isOpen) {
    // increment counter to avoid TypeError
    setInitCounter(initCounter + 1);
            
    setTimeout(() => {
      tinymce.init({
        ...
        selector: '#myTinyMCE' + (initCounter + 1), // see that for some reason I didn't investigaded, initCounter value was not updated in this particular code block
        ...
        });
        setContent("<p>Hello world</p>") // defaultValue
      }, 500) // allow some time, otherwise "real" content is not loaded in DOM
    }
  }, [isOpen]); // I don't recommend using empty array as a "componentDidMount" because is fired by the IonModal parent even without opening modal, unexpectedly
```
- later on, if needed, when submitting or whatever, you can read the content like this `tinymce.get("myTinyMCE" + initCounter).getContent();` and use `setContent()` for example

Sure this can be improved, but as a start point I hope it helps other people solving this problem without having to look out for another editor until this wrapper works for this situations.

**Comment by requefwscience** [comment]
Facing the same issue, although not with an IonModal but a pretty similar use case with an internal modal component that we use. It happens when the modal that mounts it gets unmounted and mounted for a second time.

**Comment by tiny-james** [comment]
@requefwscience does your situation use shadow dom?

**Comment by requefwscience** [comment]
@tiny-james It does not. It happens when the parent modal is unmounted and then mounted for a second time. Something like the following:

```js
// Modal.js
<Modal>
  <Editor {...props} />
</Modal>

// Page.js
...
{someBooleanValue ? <Modal /> : null}
...
```

So in this case, `someBooleanValue` has an initial value of `false`, when it changes to `true`, it mounts `Modal` and displays the `Editor` without any issues, then the modal gets closed and `someBooleanValue` changes to `false` again (therefore unmounting `Modal`), by the time it changes to `true` for a second time, we get the following error:

![Screen Shot 2021-04-13 at 10 30 02 AM](https://user-images.githubusercontent.com/62120674/114587857-4a67f180-9c43-11eb-9533-64db903ae32a.png)

The app is an SPA so it also happens if we open the modal, leave the page, come back, then try to open it again. It always fails on the second mount. The weird thing is, if we have the modal always mounted but display it via an `isOpen` prop or something similar, it does work, even when we leave the page and come back (which unmounts the modal). Something like the following instead of the above:

```js
// Modal.js
<Modal isOpen={props.isOpen}>
  <Editor {...props} />
</Modal>

// Page.js
...
<Modal isOpen={someBooleanValue} />
...
```

**Comment by requefwscience** [comment]
Tried the above with an non-minified version, getting the following:

![Screen Shot 2021-04-13 at 12 36 30 PM](https://user-images.githubusercontent.com/62120674/114603378-e2baa200-9c54-11eb-8a89-0abc205b1233.png)

**Comment by PhilMcLachlan** [comment]
I am getting the same error when opening a second time a custom popup cell editor I made for ag-grid that uses TinyMCE-react.  Any direction on how to fix this would be helpful.

**Comment by Marruixu** [comment]
Hi @PhilMcLachlan @requefwscience if you're looking for a workaround my message from above may help you :)

**Basically consists in using raw tinymce library, not the React wrapper, and changing the Editor `id` attribute each time the Modal/popup/etc is opened (with a basic counter)**

**Comment by antoniocapelo** [comment]
I'm suffering from the same issue. Although the editor is not inside a modal, it's an element that either gets added or removed from the DOM. As soon as it's added a second time, I'm faced with the "Cannot read property 'open' of undefined" error. 
@Marruixu I've tried sending a different `id` on each mount but without any luck..

**Comment by Marruixu** [comment]
Hi @antoniocapelo I've realized I wrote a mistake (fixed!) in my last reply because "changing the Editor `id` attribute" does not apply since the workaround uses the main library, not the React wrapper tinymce-react and `Editor` component becomes a `textarea` HTML element, just to clarify.

Therefore, if you have not read my previous answer, give it a try because it's a raw and easy implementation (not so simple as with the wrapper of course, but still). Maybe it works for you. Regards!

**Comment by antoniocapelo** [comment]
Thanks for the reply @Marruixu. Yes I eventually also tried using tinymce directly with a `textarea` (no react wrapper) but bumped into other problems (missing themes, icons) and even though I tried importing them, the current project's gulp setup is not ready for this approach.. I put a pin on it because of the time that this adaptation might take.

Still, it seems that  the solution is simply not using the library, which is a pitty ...

**Comment by tiny-james** [comment]
I created a replication using the react-modal component here as recommend and it works fine for me:
https://codesandbox.io/s/tinymcereact-in-modal-4bxft

I also tried hiding and showing the modal without changing the `isOpen` prop and it still works for me:
https://codesandbox.io/s/tinymcereact-unmount-modal-frrxf

It seems to work fine for me. Can someone who is getting errors show me a replication case in codesandbox (one without shadow-dom)?

**Comment by tiny-james** [comment]
Well I've traced the initial stack trace down to:
https://github.com/tinymce/tinymce/blob/develop/modules/tinymce/src/core/main/ts/init/InitContentBody.ts#L384

It seems that `editor.getDoc()` is returning `undefined` on the second call.

So I dug into the [`getDoc()`](https://github.com/tinymce/tinymce/blob/develop/modules/tinymce/src/core/main/ts/api/Editor.ts#L1021) call to see if I can figure out why that was.
```ts
  public getDoc(): Document {
    const self = this;
    let win;

    if (!self.contentDocument) {
      win = self.getWin();

      if (win) {
        self.contentDocument = win.document;
      }
    }

    return self.contentDocument;
  }
```

It calls [`getWin()`](https://github.com/tinymce/tinymce/blob/develop/modules/tinymce/src/core/main/ts/api/Editor.ts#L1000) which also returns undefined.

```ts
  public getWin(): Window {
    const self = this;
    let elm;

    if (!self.contentWindow) {
      elm = self.iframeElement;

      if (elm) {
        self.contentWindow = elm.contentWindow;
      }
    }

    return self.contentWindow;
  }
```

Digging into `getWin()` I found that the iframeElement that the editor had cached was not attached to the document.

![Selection_025](https://user-images.githubusercontent.com/41130888/118771024-5c355800-b8c5-11eb-9254-dd3a81a1111b.png)

So the question is... Why is ionmodal removing the react element without triggering an unmount?

**Comment by tiny-james** [comment]
Ok, I get it now. IonModal seems to be trying to create the modal in a node that is not in the document and then add it to the document. It works on the first load because the script tag load delays tinymce so it inits after the node is in the document, however once it has already loaded it calls init instantly the second time and tries to start on an element not in the document. This causes TinyMCE to crash when it can't get the document of the iframe which it needs to do to start (Iframes only have a document when they are rooted).

**Comment by tiny-james** [comment]
So to work-around IonModal's bad behavior I will check that I have a rooted node to initialise on before calling `tinymce.init`. I guess I will do a small amount of polling if it is unrooted and then give up.

**Comment by tiny-james** [comment]
@1dl3 I think I have a solution:
https://codesandbox.io/s/tinymce-issues204-fixed-kvw6t

I will put in a PR.

**Comment by tiny-james** [comment]
PR is here:
https://github.com/tinymce/tinymce-react/pull/264/files

**Comment by 1dl3** [comment]
@tiny-james awesome! 
Thank you for your work investigating and fixing the problem!

**Comment by tiny-james** [comment]
I have released version 3.12.5 which contains a fix for this issue at least in IonModal. I am uncertain if this will fix the issue for other modals but the stack traces certainly suggest that it is the same problem so give it a try.

If you still have problems in the other modals please open a new issue along with replication details.

**Comment by livankovic94** [comment]
I had the same problem with my UI library, which used antd ("antd": "4.24.8") in the background. The problem only appears when using the modal (second mount) + tinyMce editor ("tinymce": "5.10.2"). Interestingly, using the modal directly from antd doesn't cause the same problem.
thx @tiny-james !
